typed-racket-lib/typed-racket/base-env/unit-prims.rkt

```diff
@@ -0,0 +1,767 @@
+          "base-types-extra.rkt"
+          (for-label "colon.rkt")
+          (for-template (rep type-rep))
+          (submod "../typecheck/internal-forms.rkt" forms))
+
+(begin-for-syntax
+  (define-literal-set colon #:for-label (:))
+
+  ;; TODO: extend for other sig forms
+  (define-syntax-class def-sig-form
+    #:literal-sets (colon)
+    (pattern [name:id : type]
+             #:with internal-form #'(name type)
+             #:with erased #'name))
+
+  (define-splicing-syntax-class extends-form
+    #:literals (extends)
+    (pattern (~seq extends super:id)
+             #:with internal-form #'super
+             #:with extends-id #'super
+             #:attr form #'(extends super))
+    (pattern (~seq)
+             #:with internal-form #'#f
+             #:with extends-id '()
+             #:attr form '()))
+
+  (define-splicing-syntax-class init-depend-form
+    #:literals (init-depend)
+    (pattern (init-depend sig:id ...)
+             #:attr form (list #'(init-depend sig ...))
+             #:with names #'(sig ...))
+    (pattern (~seq)
+             #:attr form '()
+             #:with names #'()))
+  
+  (define-syntax-class unit-expr
+    (pattern e
+             #:with val #'e))
+  
+  ;; More general handling of import/export signatures in units
+  (define-syntax-class unit-imports
+    #:literals (import)
+    (pattern (import sig:sig-spec ...)
+             #:with names #'(sig.sig-name ...)
+             #:attr renamers (attribute sig.rename)))
+
+  (define-syntax-class unit-exports
+    #:literals (export)
+    (pattern (export sig:sig-spec ...)
+             #:with names #'(sig.sig-name ...)
+             #:attr renamers (attribute sig.rename)))
+  
+  (define-syntax-class sig-spec
+    #:literals (prefix rename)
+    (pattern sig-id:id
+             #:attr rename (lambda (id) id)
+             #:with sig-name #'sig-id)
+    (pattern (prefix p:id sig:sig-spec)
+             #:attr rename (lambda (id) (format-id #'sig.sig-name
+                                                 "~a~a"
+                                                 #'p
+                                                 ((attribute sig.rename) id)))
+             #:with sig-name #'sig.sig-name)
+    (pattern (rename sig:sig-spec (new:id old:id) ...)
+             #:attr rename 
+             (lambda (id)
+               (define mapping  (map cons  
+                                     (syntax->list #'(old ...))
+                                     (syntax->list #'(new ...))))
+               (define (lookup id) 
+                 (define lu (member id mapping 
+                                    (lambda (x y) 
+                                      (free-identifier=? x (car y)))))
+                 (and lu (cdar lu)))
+
+               (define rn ((attribute sig.rename) id))
+               (or (lookup rn) rn))
+             #:with sig-name #'sig.sig-name))
+  
+  ;; need lexical signature vars to error with duplicate type annotations
```

Comment:
This comment confuses me since AFAICT this function below does not emit duplicate type errors. Can you clarify it? (also in general: contracts and purpose statements for helpers)